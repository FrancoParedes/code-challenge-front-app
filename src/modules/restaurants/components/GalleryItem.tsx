import Image from 'next/image';
import Rating from '@/modules/shared/components/Rating';
import React from 'react';
import { IRestaurant } from '@/modules/shared/interfaces/IRestaurant';
import Skeleton from 'react-loading-skeleton';
import 'react-loading-skeleton/dist/skeleton.css';

type IPartialRestaurant = Pick<
  IRestaurant,
  'url' | 'alias' | 'id' | 'name' | 'rating' | 'price' | 'url' | 'image_url'
>;

interface IGalleryItemProps {
  restaurant: IPartialRestaurant | null;
  loading?: boolean;
}

const validateFullData = (restaurant: IPartialRestaurant) => {
  const { id, image_url, alias, name, rating, url } = restaurant;
  return !!(id && image_url && alias && name && rating && url);
};

const GalleryItem: React.FC<IGalleryItemProps> = ({ restaurant, loading = false }) => {
  const hasAllInformation = restaurant && validateFullData(restaurant);
  if (!hasAllInformation && !loading) {
    return null;
  }

  const handleOnClick = (e: React.MouseEvent) => {
    e.stopPropagation();
    window.open(restaurant?.url, '_blank');
  };

  return (
    <div className={`gallery-item ${!loading ? 'border' : ''}`} onClick={handleOnClick}>
      <div className="gallery-item-image">
        {!loading && restaurant && (
          <Image
            src={restaurant?.image_url}
            alt={restaurant?.alias}
            width="105"
            height="105"
          />
        )}
        {loading && <Skeleton width={105} height={105} />}
      </div>
      <div className="gallery-item-body">
        <div className="gallery-item-title">
          {restaurant?.name || <Skeleton width="100%" />}
        </div>
        <div className="gallery-item-footer">
          <div>
            {restaurant?.rating ? (
              <Rating rating={restaurant.rating} />
            ) : (
              <Skeleton width={80} />
            )}
          </div>
          {restaurant?.price && <div>{restaurant.price}</div>}
          {loading && <Skeleton width={50} />}
        </div>
        <div className="gallery-item-button">
          {restaurant && (
            <button type="button" onClick={handleOnClick}>
              VIEW
            </button>
          )}
          {!restaurant && <Skeleton width={100} height={30} />}
        </div>
      </div>
    </div>
  );
};

export default GalleryItem;
