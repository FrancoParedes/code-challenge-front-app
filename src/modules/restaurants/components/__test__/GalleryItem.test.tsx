import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import GalleryItem from '@/modules/restaurants/components/GalleryItem';

const mockRestaurant = {
  id: '1',
  image_url: 'https://s3-media3.fl.yelpcdn.com/bphoto/cz4PKevCEGTlutQZm6i2BQ/o.jpg',
  alias: 'restaurant-alias',
  name: 'Restaurant Name',
  rating: 4.5,
  price: '$$',
  url: 'https://example.com/restaurant',
};

window.open = jest.fn();

describe('test Gallery item', () => {
  it('render with complete data', () => {
    render(<GalleryItem restaurant={mockRestaurant} />);

    const image = screen.getByAltText(mockRestaurant.alias);
    expect(image).toBeInTheDocument();

    expect(screen.getByText(mockRestaurant.name)).toBeInTheDocument();
    expect(screen.getByText(mockRestaurant.price)).toBeInTheDocument();

    expect(screen.getByText('VIEW')).toBeInTheDocument();
  });

  it('render incomplete data', () => {
    const { container } = render(<GalleryItem restaurant={null} />);
    expect(container.firstChild).toBeNull();
  });

  it('click item', () => {
    render(<GalleryItem restaurant={mockRestaurant} />);
    const buttonItem = screen.getByText('VIEW');
    fireEvent.click(buttonItem);
    expect(window.open).toHaveBeenCalledWith(mockRestaurant.url, '_blank');
  });

  it('render loading', () => {
    const { container } = render(<GalleryItem restaurant={null} loading />);
    expect(container).toMatchSnapshot();
  });
});
