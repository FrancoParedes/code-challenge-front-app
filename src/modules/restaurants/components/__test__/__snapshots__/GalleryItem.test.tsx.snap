// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Valida el clic en el componente GalleryItem 1`] = `
<div>
  <div
    class="gallery-item "
  >
    <div
      class="gallery-item-image"
    >
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="width: 105px; height: 105px;"
        >
          ‌
        </span>
        <br />
      </span>
    </div>
    <div
      class="gallery-item-body"
    >
      <div
        class="gallery-item-title"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 100%;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="gallery-item-footer"
      >
        <div>
          <span
            aria-busy="true"
            aria-live="polite"
          >
            <span
              class="react-loading-skeleton"
              style="width: 80px;"
            >
              ‌
            </span>
            <br />
          </span>
        </div>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 50px;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="gallery-item-button"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 100px; height: 30px;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`test Gallery item render loading 1`] = `
<div>
  <div
    class="gallery-item "
  >
    <div
      class="gallery-item-image"
    >
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
          style="width: 105px; height: 105px;"
        >
          ‌
        </span>
        <br />
      </span>
    </div>
    <div
      class="gallery-item-body"
    >
      <div
        class="gallery-item-title"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 100%;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="gallery-item-footer"
      >
        <div>
          <span
            aria-busy="true"
            aria-live="polite"
          >
            <span
              class="react-loading-skeleton"
              style="width: 80px;"
            >
              ‌
            </span>
            <br />
          </span>
        </div>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 50px;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="gallery-item-button"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 100px; height: 30px;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
    </div>
  </div>
</div>
`;
